File: src/secretd/redact.py
"""Streaming, encoding-aware redaction of secret values.

The redactor sits between the child process's PTY and the response that goes
back to the agent.  It has to work on a stream (output arrives in arbitrary
chunks) and it has to catch a value even when the program that printed it
mangled the bytes on the way out -- colour codes spliced into the middle,
base64 with line wrapping, URL escaping, shell quoting.

See ``docs/redaction.md`` for the reasoning behind each stage.
"""

from __future__ import annotations

import base64
import json
import math
import re
import shlex
import urllib.parse
from collections import Counter
from dataclasses import dataclass
from typing import Sequence

# --------------------------------------------------------------------------
# Stage 1: ANSI / control-character stripping
# --------------------------------------------------------------------------

_ANSI_RE = re.compile(
    "|".join(
        [
            r"\x1b\][^\x07\x1b]*(?:\x07|\x1b\\)",  # OSC ... BEL / ST
            r"\x1b[P^_X][^\x1b]*\x1b\\",  # DCS / PM / APC / SOS
            r"\x1b\[[0-?]*[ -/]*[@-~]",  # CSI
            r"\x1b[()][B0UK]",  # charset selection
            r"\x1b[@-Z\\-_]",  # two-character escapes
            r"[\x00-\x08\x0b\x0c\x0e-\x1a\x1c-\x1f\x7f]",  # stray controls
        ]
    )
)

# How far back an incomplete escape sequence may reasonably start.
_MAX_ESCAPE_LEN = 64


def strip_ansi(text: str) -> str:
    """Remove escape sequences and normalise CRLF.  Not stream-safe on its own."""
    return _ANSI_RE.sub("", text).replace("\r\n", "\n")


def _strip_ansi_stream(buf: str) -> tuple[str, str]:
    """Strip escapes from ``buf``, holding back a possibly-incomplete tail.

    Returns ``(clean, carry)``.  ``carry`` is text that must be prepended to the
    next chunk because it may be the beginning of an escape sequence (or a lone
    ``\\r`` that could turn out to be the first half of a CRLF).
    """
    carry_start = len(buf)
    esc = buf.rfind("\x1b")
    if esc != -1 and len(buf) - esc <= _MAX_ESCAPE_LEN:
        # Only hold back if the sequence is not obviously already terminated.
        if not _ANSI_RE.match(buf, esc):
            carry_start = esc
    if carry_start == len(buf) and buf.endswith("\r"):
        carry_start = len(buf) - 1
    head, carry = buf[:carry_start], buf[carry_start:]
    return strip_ansi(head), carry


# --------------------------------------------------------------------------
# Stage 2: the expanded value set
# --------------------------------------------------------------------------


def base64_variants(value: str) -> set[str]:
    """Base64 encodings of ``value``: standard and URL-safe, padded and not."""
    raw = value.encode("utf-8", "surrogatepass")
    out: set[str] = set()
    for enc in (base64.b64encode(raw), base64.urlsafe_b64encode(raw)):
        text = enc.decode("ascii")
        out.add(text)
        out.add(text.rstrip("="))
    return out


def variants(value: str) -> set[str]:
    """Every rendering of ``value`` the redactor knows how to recognise.

    Deliberately *not* exhaustive -- an agent that wants to defeat this can
    (see the threat model).  These are the encodings that ordinary tools
    produce by accident: JSON output, URLs, ``set -x`` traces, base64 dumps.
    """
    out: set[str] = {value}
    out |= base64_variants(value)
    out.add(urllib.parse.quote(value, safe=""))
    out.add(urllib.parse.quote_plus(value))
    out.add(json.dumps(value)[1:-1])  # JSON string escaping, without the quotes
    out.add(shlex.quote(value))  # 'value' or value
    out.add(value.replace("'", "'\\''"))  # body of a shell single-quoted string
    out.add(
        value.replace("\\", "\\\\")
        .replace('"', '\\"')
        .replace("$", "\\$")
        .replace("`", "\\`")
    )  # body of a shell double-quoted string
    return {v for v in out if v}


# --------------------------------------------------------------------------
# Stage 3: eligibility -- refuse to redact values that would eat the output
# --------------------------------------------------------------------------


def shannon_entropy(value: str) -> float:
    """Shannon entropy in bits per character."""
    if not value:
        return 0.0
    counts = Counter(value)
    n = len(value)
    return -sum((c / n) * math.log2(c / n) for c in counts.values())


@dataclass(frozen=True)
class EligibilityPolicy:
    min_length: int = 8
    min_unique_chars: int = 4
    min_entropy_bits_per_char: float = 1.5

    def check(self, value: str) -> str | None:
        """Return ``None`` if the value may be redacted, else a reason string."""
        if len(value) < self.min_length:
            return f"shorter than {self.min_length} characters"
        if len(set(value)) < self.min_unique_chars:
            return f"fewer than {self.min_unique_chars} distinct characters"
        entropy = shannon_entropy(value)
        if entropy < self.min_entropy_bits_per_char:
            return f"low entropy ({entropy:.2f} bits/char)"
        return None


def token_for(ref: str) -> str:
    """The stable placeholder a secret is replaced with.

    Stable across turns and across processes so the model can reason about
    "the router password" without ever seeing it.
    """
    return f"«SECRET:{ref}»"


# --------------------------------------------------------------------------
# Stage 4: the streaming redactor
# --------------------------------------------------------------------------


@dataclass
class _Entry:
    ref: str
    token: str
    pattern: re.Pattern[str]
    wrapped: re.Pattern[str] | None
    longest: int


class Redactor:
    """Replaces every known secret rendering with a stable token.

    Usage::

        r = Redactor([("home/router/admin", "hunter2hunter2")])
        out = r.feed(chunk) + ... + r.flush()
        r.counts  # {token: occurrences}

    ``feed`` withholds a tail of the stream so a value split across two reads
    is still caught; ``flush`` releases it.
    """

    def __init__(
        self,
        secrets: Sequence[tuple[str, str]],
        policy: EligibilityPolicy | None = None,
    ) -> None:
        self.policy = policy or EligibilityPolicy()
        self.skipped: list[tuple[str, str]] = []
        self.counts: Counter[str] = Counter()
        self._entries: list[_Entry] = []
        self._ansi_carry = ""
        self._buf = ""

        seen: set[str] = set()
        for ref, value in secrets:
            if not isinstance(value, str) or value in seen:
                continue
            reason = self.policy.check(value)
            if reason is not None:
                self.skipped.append((ref, reason))
                continue
            seen.add(value)
            self._entries.append(self._compile(ref, value))

        # Longest value first: if one secret is a substring of another, the
        # longer token must win.
        self._entries.sort(key=lambda e: e.longest, reverse=True)
        longest = max((e.longest for e in self._entries), default=0)
        # x2 covers base64 line wrapping (newlines inserted inside a value),
        # +16 covers quoting expansion at a chunk boundary.
        self.overlap = longest * 2 + 16

    @staticmethod
    def _compile(ref: str, value: str) -> _Entry:
        vs = sorted(variants(value), key=len, reverse=True)
        pattern = re.compile("|".join(re.escape(v) for v in vs))
        b64 = sorted(base64_variants(value), key=len, reverse=True)
        wrapped = re.compile("|".join(re.escape(v) for v in b64)) if b64 else None
        return _Entry(ref, token_for(ref), pattern, wrapped, max(len(v) for v in vs))

    # -- public API --------------------------------------------------------

    @property
    def active(self) -> bool:
        return bool(self._entries)

    def feed(self, text: str) -> str:
        """Absorb a chunk of raw output; return the part that is safe to emit."""
        if not text:
            return ""
        self._ansi_carry += text
        clean, self._ansi_carry = _strip_ansi_stream(self._ansi_carry)
        self._buf = self._redact(self._buf + clean)
        if len(self._buf) > self.overlap:
            out, self._buf = self._buf[: -self.overlap], self._buf[-self.overlap :]
            return out
        return ""

    def flush(self) -> str:
        """Release everything held back.  Call once, at end of stream."""
        tail = strip_ansi(self._ansi_carry)
        self._ansi_carry = ""
        out = self._redact(self._buf + tail)
        self._buf = ""
        return out

    def redact_text(self, text: str) -> str:
        """One-shot convenience for text that is already complete."""
        return self.feed(text) + self.flush()

    def summary(self) -> list[dict[str, object]]:
        """``redactions`` field of the wire response: tokens and counts, no values."""
        return [
            {"token": token, "count": count}
            for token, count in sorted(self.counts.items())
            if count
        ]

    # -- internals ---------------------------------------------------------

    def _redact(self, text: str) -> str:
        if not text:
            return text
        for entry in self._entries:
            if entry.wrapped is not None and ("\n" in text or "\r" in text):
                text = self._sub_wrapped(text, entry)
            text, n = entry.pattern.subn(entry.token, text)
            if n:
                self.counts[entry.token] += n
        return text

    def _sub_wrapped(self, text: str, entry: _Entry) -> str:
        """Catch base64 that has been line-wrapped (``base64`` wraps at 76 cols).

        Matching happens on a copy of the haystack with newlines removed; hits
        are mapped back to spans in the original text so the surrounding output
        is preserved.
        """
        assert entry.wrapped is not None
        keep: list[str] = []
        index: list[int] = []
        for i, ch in enumerate(text):
            if ch not in "\n\r":
                keep.append(ch)
                index.append(i)
        if len(keep) == len(text):
            return text  # no newlines to collapse; the plain pass handles it
        view = "".join(keep)

        spans: list[tuple[int, int]] = []
        for m in entry.wrapped.finditer(view):
            start, end = index[m.start()], index[m.end() - 1] + 1
            if "\n" in text[start:end] or "\r" in text[start:end]:
                spans.append((start, end))
        if not spans:
            return text

        pieces: list[str] = []
        cursor = 0
        for start, end in spans:
            if start < cursor:  # overlapping match, already covered
                continue
            pieces.append(text[cursor:start])
            pieces.append(entry.token)
            self.counts[entry.token] += 1
            cursor = end
        pieces.append(text[cursor:])
        return "".join(pieces)

File: internal/redact/redact.go
// Package redact provides streaming, encoding-aware redaction of secret values.
// See docs/redaction.md.
//
// It works on a stream, output arriving in arbitrary chunks, and has to catch a
// value the printing program mangled on the way out: colour codes spliced in,
// base64 line wrapping, URL escaping, shell quoting.
//
// Everything index-sensitive works on []rune, a byte-offset slice being able to
// cut a multi-byte character in half.
//
// Four stages, a file each, and each exists for a case the stage itself does
// not make obvious:
//
//  1. ansi.go strips ANSI and the control characters, and maps what is found in
//     the stripped text back onto the original.
//  2. variants.go expands one value into every spelling it leaves a program in:
//     base64, base32, percent, shell quoting, JSON escaping.
//  3. eligibility.go refuses a value too short to search for, which would match
//     inside ordinary words and blank unrelated output.
//  4. This file is the redactor over the result. values.go builds the set of
//     renderings it searches for, and stream.go is the buffer that holds a
//     chunk back while a rendering split across two could still be arriving.
package redact

import (
	"sort"
	"strings"
	"unicode/utf8"
)

// rendering is one spelling of one value, and the token it stands for.
type rendering struct {
	text  string
	token string
}

// Count is one row of the wire response's "redactions" field.
type Count struct {
	Token string `json:"token"`
	Count int    `json:"count"`
}

// Redactor replaces every known secret rendering with a stable token. Feed
// withholds a tail so a value split across two reads is still caught; Flush
// releases it.
//
// One per request: it carries the counts and the stream's own buffers. The set
// it matches against is shared.
type Redactor struct {
	*Values

	counts map[string]int
	// raw is the input held back for the next chunk: an incomplete escape or CRLF
	// half at the very end, and the overlap window that a value split across the
	// boundary is caught in. Held raw rather than stripped, so an escape that ate
	// a value's first byte is re-stripped with the rest of the value on the next
	// chunk instead of leaving its stripped remainder to go out in the clear.
	raw []rune
	// rawBytes holds a chunk's trailing bytes when they begin a multibyte rune
	// the chunk cut short, so the rune is decoded whole once its rest arrives
	// rather than corrupted to U+FFFD on each side of the split. Prepended to the
	// next chunk; flushed as-is at end of stream.
	rawBytes []byte
	// Bytes that were not valid UTF-8, counted rather than acted on; see Feed.
	invalidBytes int
}

// Secret is one (ref, value) pair fed to the redactor.
type Secret struct {
	Ref   string
	Value string
}

// New builds a redactor over the given secrets. A value the policy refuses is
// not matched; naming it is the secretstore package's job.
//
// A caller that redacts more than once against the same set should build the
// set with NewValues and take a Redactor per request from it.
func New(secrets []Secret, policy EligibilityPolicy) *Redactor {
	return NewValues(secrets, policy).Redactor()
}

// RedactText is a one-shot convenience for text that is already complete.
func (r *Redactor) RedactText(text string) string { return r.Feed(text) + r.Flush() }

// InvalidBytes is how many bytes of everything fed in were not valid UTF-8.
// Non-zero means the output was not text and what came back is not what the
// command wrote: an invalid byte becomes U+FFFD, and the C0 controls that fill
// binary are stripped outright. A caller that pipes the output somewhere
// reports this, so a corrupted archive is visible when it is produced.
func (r *Redactor) InvalidBytes() int { return r.invalidBytes }

// invalidUTF8Bytes counts the bytes in text that do not begin a valid rune.
func invalidUTF8Bytes(text string) int {
	n := 0
	for i := 0; i < len(text); {
		r, size := utf8.DecodeRuneInString(text[i:])
		if r == utf8.RuneError && size == 1 {
			n++
		}
		i += size
	}
	return n
}

// Summary is the wire response's "redactions": tokens and counts, never values.
func (r *Redactor) Summary() []Count {
	out := []Count{}
	for token, count := range r.counts {
		if count > 0 {
			out = append(out, Count{Token: token, Count: count})
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Token < out[j].Token })
	return out
}

// tokenSpan is a match to replace: the offsets it covers and the token it
// stands for. The offsets are into whatever source the splice reads, byte or
// rune, so long as size and between below share that index space.
type tokenSpan struct {
	start, end int
	token      string
}

// spliceSpans writes source with each span replaced by its token. size is the
// source length in the spans' index space, and between returns source[a:b] in
// it. A span that starts inside the previous one covers only its uncovered tail,
// and one wholly covered is skipped, so partially overlapping secrets are both
// redacted without a negative slice.
func (r *Redactor) spliceSpans(spans []tokenSpan, size int, between func(a, b int) string) string {
	var b strings.Builder
	b.Grow(size)
	cursor := 0
	for _, s := range spans {
		if s.end <= cursor {
			continue
		}
		if s.start < cursor {
			b.WriteString(s.token)
		} else {
			b.WriteString(between(cursor, s.start))
			b.WriteString(s.token)
		}
		r.counts[s.token]++
		cursor = s.end
	}
	b.WriteString(between(cursor, size))
	return b.String()
}

func (r *Redactor) redact(text string, ev *escapeView) string {
	if text == "" || r.matcher == nil {
		return text
	}
	// The escape pass first, against a view holding the byte a CSI took from the
	// front of a value. Only what that view alone can find: everything still
	// contiguous in text is left to the two passes below, so the ordinary case
	// pays nothing here and this one is not scanned twice.
	if out, changed := r.subEscaped(text, ev); changed {
		text = out
	}
	// The wrapped pass next, against a newline-free view of the output, so it
	// catches a rendering a formatter split across lines: base64 wraps at 76
	// columns, and `fold` wraps every rendering the same way. The newline guard
	// in subWrapped keeps it to genuinely line-spanning matches, so the plain
	// pass below still owns everything on a single line.
	//
	// Newlines only: a continuation the formatter indents still has whitespace
	// between the fragments and is not caught. Collapsing the indentation too
	// would join any two words straddling an indented line break, which corrupts
	// more output than the wrapping it would catch.
	if out, changed := r.subWrapped(newCollapsedView(text)); changed {
		text = out
	}
	// One pass, whatever the number of values: the automaton carries every
	// rendering and the match says which value it was.
	spans := r.matcher.find(text)
	if len(spans) == 0 {
		// Nothing built and nothing copied, which is the ordinary case: most output
		// carries no secret at all.
		return text
	}
	toks := make([]tokenSpan, 0, len(spans))
	for _, s := range spans {
		token, ok := r.tokenOf[text[s.start:s.end]]
		if !ok {
			continue // unreachable: the automaton is built from these keys
		}
		toks = append(toks, tokenSpan{start: s.start, end: s.end, token: token})
	}
	return r.spliceSpans(toks, len(text), func(a, b int) string { return text[a:b] })
}

// collapsedView is one haystack with its line breaks taken out, plus what maps
// a match back onto the original: a formatter can wrap a value across lines, so
// matching happens against view and the replaced span is in the original.
type collapsedView struct {
	// runes is the original, indexed the way the spans below are.
	runes []rune
	view  string
	// byteStart maps a byte offset in view to the index in runes of the rune
	// beginning there, plus one entry for the end.
	byteStart []int
	// collapsed is false when there were no line breaks, in which case the plain
	// pass covers everything this would find.
	collapsed bool
}

func newCollapsedView(text string) *collapsedView {
	v := &collapsedView{}
	if !strings.ContainsAny(text, "\n\r") {
		return v
	}
	v.collapsed = true
	v.runes = []rune(text)
	var b strings.Builder
	b.Grow(len(text))
	v.byteStart = make([]int, 0, len(text)+1)
	for i, ch := range v.runes {
		if ch == '\n' || ch == '\r' {
			continue
		}
		for range utf8.RuneLen(ch) {
			v.byteStart = append(v.byteStart, i)
		}
		b.WriteRune(ch)
	}
	v.byteStart = append(v.byteStart, len(v.runes))
	v.view = b.String()
	return v
}

// subEscaped replaces every value that is whole only in the escape view, and
// reports whether it changed anything. The span it replaces is in text: the
// byte the sequence took is not there, so the token stands where what survived
// of the value stood.
func (r *Redactor) subEscaped(text string, v *escapeView) (string, bool) {
	if v == nil || !v.lenient {
		return "", false
	}
	var spans []tokenSpan
	for _, loc := range r.matcher.find(v.view) {
		token, ok := r.tokenOf[v.view[loc.start:loc.end]]
		if !ok {
			continue // unreachable: the automaton is built from these keys
		}
		start, end := v.clean[loc.start], v.clean[loc.end]
		// Only a match a returned byte made whole. One the same length in both is
		// contiguous in text as well, and the plain pass owns it: taking it here
		// would count it and then leave that pass nothing, which reads the same
		// but pays for two scans of every value.
		if end-start == loc.end-loc.start || end <= start {
			continue
		}
		spans = append(spans, tokenSpan{start: start, end: end, token: token})
	}
	if len(spans) == 0 {
		return "", false
	}
	return r.spliceSpans(spans, len(text), func(a, b int) string { return text[a:b] }), true
}

// subWrapped replaces every line-wrapped rendering, and reports whether it
// changed anything. Its spans are offsets into the original runes, so the splice
// reads v.runes rather than a byte string.
func (r *Redactor) subWrapped(v *collapsedView) (string, bool) {
	if !v.collapsed {
		return "", false
	}
	var spans []tokenSpan
	for _, loc := range r.matcher.find(v.view) {
		token, ok := r.tokenOf[v.view[loc.start:loc.end]]
		if !ok {
			continue // unreachable: the automaton is built from these keys
		}
		// byteStart is indexed by byte, so the end comes from the match's last byte
		// rather than the offset after it.
		start, end := v.byteStart[loc.start], v.byteStart[loc.end-1]+1
		if strings.ContainsAny(string(v.runes[start:end]), "\n\r") {
			spans = append(spans, tokenSpan{start: start, end: end, token: token})
		}
	}
	if len(spans) == 0 {
		return "", false
	}
	return r.spliceSpans(spans, len(v.runes), func(a, b int) string {
		return string(v.runes[a:b])
	}), true
}
